kolibri/core/deviceadmin/utils.py

```diff
@@ -0,0 +1,159 @@
+    pass
+
+
+def default_backup_folder():
+    return os.path.join(os.environ['KOLIBRI_HOME'], 'backups')
+
+
+def get_dtm_from_backup_name(fname):
+    """
+    Returns the date time string from our automated backup filenames
+    """
+    p = re.compile(r"^db\-v[^_]+_(?P<dtm>[\d\-_]+).*\.dump$")
+    m = p.search(fname)
+    if m:
+        return m.groups("dtm")[0]
+    raise ValueError(
+        "Tried to get date component of unparsed filename: {}".format(fname)
+    )
+
+
+def is_full_version(fname):
+    """
+    Tells us if a backup file name is named as if it's from the exact same
+    version.
+
+    Supposes versions do not contain underscores '_'
+    """
+    # Can contain suffixes denoting alpha, beta, post, dev etc.
+    full_version = kolibri.__version__
+    return fname.startswith(
+        "db-v{}_".format(full_version)
+    )
+
+
+def dbbackup(old_version, dest_folder=None):
+    """
+    Sqlite3 only
+
+    Backup database to dest_folder. Uses SQLite's built in iterdump():
+    https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.iterdump
+
+    Notice that it's important to add at least version and date to the path
+    of the backup, otherwise you risk that upgrade activities carried out on
+    the same date overwrite each other. It's also quite important for the user
+    to know which version of Kolibri that a certain database should match.
+
+    :param: dest_folder: Default is ~/.kolibri/backups/db-[version]-[date].dump
+
+    :returns: Path of new backup file
+    """
+
+    if 'sqlite3' not in settings.DATABASES['default']['ENGINE']:
+        raise IncompatibleDatabase()
+
+    if not dest_folder:
+        dest_folder = default_backup_folder()
+
+    # This file name is a convention, used to figure out the latest backup
+    # that was made (by the dbrestore command)
+    fname = "db-v{version}_{dtm}.dump".format(
+        version=old_version,
+        dtm=datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
+    )
+
+    if not os.path.exists(dest_folder):
+        os.makedirs(dest_folder)
+
+    backup_path = os.path.join(dest_folder, fname)
+
+    with open(backup_path, 'w') as f:
+        # If the connection hasn't been opened yet, then open it
+        if not db.connections['default'].connection:
+            db.connections['default'].connect()
+        for line in db.connections['default'].connection.iterdump():
+            f.write(line)
+
+    return backup_path
+
+
+def dbrestore(from_file):
+    """
+    Sqlite3 only
+
+    Restores the database given a special database dump file containing SQL
+    statements.
+    """
+
+    if 'sqlite3' not in settings.DATABASES['default']['ENGINE']:
+        raise IncompatibleDatabase()
+
+    dst_file = settings.DATABASES['default']['NAME']
+
+    # Close connections
+    db.connections.close_all()
+
+    # Wipe current database file
+    if not db.connections['default'].is_in_memory_db(dst_file):
+        open(dst_file, "w").truncate()
```

Comment:
Shouldn't this be a `with`, to make sure it gets closed after?

Reply:
I think Python garbage collection is cool with this :) There's no reference to it, so should be closed immediately - I use this pattern often

Reply:
I couldn't find a proper source, so better safe than sorry and changed it! Thanks :)